File: Module/PlayWays/src/main/java-room-grab/com/module/playways/grab/room/view/GrabRootView.java
package com.module.playways.grab.room.view;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import com.common.view.ex.ExRelativeLayout;

import java.util.ArrayList;
import java.util.List;

public class GrabRootView extends ExRelativeLayout {
    List<OnTouchListener> mOnTouchListeners = new ArrayList<>();

    public GrabRootView(Context context) {
        super(context);
        init();
    }

    public GrabRootView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public GrabRootView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    void init() {
        setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                for (OnTouchListener l : mOnTouchListeners) {
                    boolean r = l.onTouch(v, event);
                    if (r) {
                        return true;
                    }
                }
                return false;
            }
        });
    }

    public void addOnTouchListener(OnTouchListener l) {
        if (mOnTouchListeners.contains(l)) {
            mOnTouchListeners.remove(l);
        }
        mOnTouchListeners.add(l);
    }

    public void removeOnTouchListener(OnTouchListener l) {
        mOnTouchListeners.remove(l);
    }
}

File: Module/PlayWays/src/main/java-room-grab/com/module/playways/grab/room/view/GrabRootView.kt
package com.module.playways.grab.room.view

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

import com.common.view.ex.ExRelativeLayout

import java.util.ArrayList

class GrabRootView : ExRelativeLayout {

    private var mOnTouchListeners: MutableList<View.OnTouchListener> = ArrayList()

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    init {
        setOnTouchListener(OnTouchListener { v, event ->
            for (l in mOnTouchListeners) {
                val r = l.onTouch(v, event)
                if (r) {
                    return@OnTouchListener true
                }
            }
            false
        })
    }

    fun addOnTouchListener(l: View.OnTouchListener) {
        if (mOnTouchListeners.contains(l)) {
            mOnTouchListeners.remove(l)
        }
        mOnTouchListeners.add(l)
    }

    fun removeOnTouchListener(l: View.OnTouchListener) {
        mOnTouchListeners.remove(l)
    }
}
